pkg/internal/cluster/providers/docker/provider.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or impliep.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package docker
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"sigs.k8s.io/kind/pkg/cluster/constants"
+	"sigs.k8s.io/kind/pkg/cluster/nodes"
+	"sigs.k8s.io/kind/pkg/errors"
+	"sigs.k8s.io/kind/pkg/exec"
+
+	"sigs.k8s.io/kind/pkg/cluster/nodeutils"
+	"sigs.k8s.io/kind/pkg/internal/apis/config"
+	"sigs.k8s.io/kind/pkg/internal/cluster/providers/provider"
+	"sigs.k8s.io/kind/pkg/internal/cluster/providers/provider/common"
+	"sigs.k8s.io/kind/pkg/internal/util/cli"
+)
+
+// NewProvider returns a new provider based on executing `docker ...`
+func NewProvider() provider.Provider {
+	return &Provider{}
+}
+
+// Provider implements provider.Provider
+// see NewProvider
+type Provider struct{}
+
+// Provision is part of the providers.Provider interface
+func (p *Provider) Provision(status *cli.Status, cluster string, cfg *config.Cluster) (err error) {
+	// TODO: validate cfg
+	// ensure node images are pulled before actually provisioning
+	ensureNodeImages(status, cfg)
+
+	// actually provision the cluster
+	// TODO: strings.Repeat("📦", len(desiredNodes))
+	status.Start("Preparing nodes 📦")
+	defer func() { status.End(err == nil) }()
+	return provision(cluster, cfg)
+}
+
+// ListClusters is part of the providers.Provider interface
+func (p *Provider) ListClusters() ([]string, error) {
+	cmd := exec.Command("docker",
+		"ps",
+		"-q",         // quiet output for parsing
+		"-a",         // show stopped nodes
+		"--no-trunc", // don't truncate
+		// filter for nodes with the cluster label
+		"--filter", "label="+constants.ClusterLabelKey,
+		// format to include the cluster name
+		"--format", fmt.Sprintf(`{{.Label "%s"}}`, constants.ClusterLabelKey),
+	)
+	lines, err := exec.OutputLines(cmd)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list clusters")
+	}
+	return lines, nil
```

Comment:
won't this repeat the cluster name for every node in multi node clusters? Probably dedup this in a sets.String

Reply:
Ack. This is a bug.

Reply:
fixed